crates/polars-stream/src/nodes/io_sinks/csv.rs

```diff
@@ -0,0 +1,164 @@
+use polars_io::prelude::{CsvWriter, CsvWriterOptions};
+use polars_io::SerWriter;
+use polars_utils::priority::Priority;
+
+use crate::async_primitives::linearizer::Linearizer;
+use crate::nodes::{ComputeNode, JoinHandle, MorselSeq, PortState, TaskPriority, TaskScope};
+use crate::pipe::{RecvPort, SendPort};
+use crate::DEFAULT_LINEARIZER_BUFFER_SIZE;
+
+pub struct CsvSinkNode {
+    path: PathBuf,
+    schema: SchemaRef,
+
+    write_options: CsvWriterOptions,
+}
+
+impl CsvSinkNode {
+    pub fn new(
+        schema: SchemaRef,
+        path: &Path,
+        write_options: &CsvWriterOptions,
+    ) -> PolarsResult<Self> {
+        Ok(Self {
+            path: path.to_path_buf(),
+            schema,
+
+            write_options: write_options.clone(),
+        })
+    }
+}
+
+impl ComputeNode for CsvSinkNode {
+    fn name(&self) -> &str {
+        "csv_sink"
+    }
+
+    fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) -> PolarsResult<()> {
+        assert!(send.is_empty());
+        assert!(recv.len() == 1);
+
+        // We are always ready to receive, unless the sender is done, then we're
+        // also done.
+        if recv[0] != PortState::Done {
+            recv[0] = PortState::Ready;
+        }
+
+        Ok(())
+    }
+
+    fn spawn<'env, 's>(
+        &'env mut self,
+        scope: &'s TaskScope<'s, 'env>,
+        recv_ports: &mut [Option<RecvPort<'_>>],
+        send_ports: &mut [Option<SendPort<'_>>],
+        _state: &'s ExecutionState,
+        join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+    ) {
+        assert!(recv_ports.len() == 1);
+        assert!(send_ports.is_empty());
+
+        // .. -> Encode task
+        let receivers = recv_ports[0].take().unwrap().parallel();
+        // Encode tasks -> IO task
+        let (mut linearizer, senders) = Linearizer::<Priority<Reverse<MorselSeq>, Vec<u8>>>::new(
+            receivers.len(),
+            DEFAULT_LINEARIZER_BUFFER_SIZE,
+        );
+
+        let slf = &*self;
+
+        // 16MB
+        const DEFAULT_ALLOCATION_SIZE: usize = 1 << 24;
+
+        // Encode task.
+        //
+        // Task encodes the columns into their corresponding CSV encoding.
+        for (mut receiver, mut sender) in receivers.into_iter().zip(senders) {
+            join_handles.push(scope.spawn_task(TaskPriority::High, async move {
+                // Amortize the allocations over time. If we see that we need to do way larger
+                // allocations, we adjust to that over time.
+                let mut allocation_size = DEFAULT_ALLOCATION_SIZE;
+                let options = slf.write_options.clone();
+
+                while let Ok(morsel) = receiver.recv().await {
+                    let (df, seq, _, _) = morsel.into_inner();
+
+                    let mut buffer = Vec::with_capacity(allocation_size);
+                    let mut writer = CsvWriter::new(&mut buffer)
+                        .include_bom(false) // Handled once in the IO task.
```

Comment:
Maybe we can hoist the writer out of the while loop?

Reply:
I don't think it matters much since the `CsvWriter` creation is cheap, and we cannot reuse the allocation anyway because it needs to be passed to the IO task.